### ISSUE TITLE
Support dual-stack when `"subnet": "usePodCidr",`

### ISSUE THREAD

**Issue by manuelbuil** [projectcalico/cni-plugin#1177]
## Expected Behavior

When using a dual-stack pod-cidr, pods should have two ips in the network interface, one ipv4 and one ipv6

## Current Behavior

When using a dual-stack pod-cidr, pods have only one ipv4 address

## Possible Solution

Fetch node.Spec.PodCIDRs instead of node.Spec.PodCIDR in https://github.com/projectcalico/cni-plugin/blob/v3.19.2/pkg/k8s/k8s.go#L894-L909 and then adapt the code https://github.com/projectcalico/cni-plugin/blob/v3.19.2/internal/pkg/utils/utils.go#L332-L340 to process a slice instead of a string 

## Steps to Reproduce (for bugs)

1. Deploy k8s with a dual-stack pod-cidr (e.g. cluster-cidr: 10.42.0.0/16,2001:cafe:42:0::/56)
2. Use canal with the ipam config:
              "type": "host-local",
              "subnet": "usePodCidr"

## Your Environment

* CNI plugin version: 3.19.2
* Orchestrator version: Kubernetes 1.21.5
* Operating System and version: Ubuntu
* Link to your project (optional): https://github.com/rancher/rke2/issues/1883

**Comment by lwr20** [comment]
Canal is flannel CNI and Calico for policy.  
This issue suggests that Flannel does not support IPv6: https://github.com/flannel-io/flannel/issues/248

**Comment by lwr20** [comment]
Ah interesting, I see that the author of this issue, added dual stack support to flannel too https://github.com/flannel-io/flannel/pull/1448

But I'd expect that the flannel part of Canal would be the part that would assign the IPs to the pod?

**Comment by manuelbuil** [comment]
> Canal is flannel CNI and Calico for policy. This issue suggests that Flannel does not support IPv6: [flannel-io/flannel#248](https://github.com/flannel-io/flannel/issues/248)

Not exactly. Canal is flannel CNI for inter-node communication and Calico for policy and intra-node communication. As part of that intra-node communication, it takes care of distributing IPs among pods using these code: https://github.com/projectcalico/cni-plugin/blob/v3.19.2/internal/pkg/utils/utils.go#L330-L342. Because it reads `PodCidr` instead of `PodCidrs` variable, it will only get one IP and thus it will not work in dual-stack. This is how the spec of a node is in a dual-stack deployment:
```
spec:
  podCIDR: 10.42.0.0/24
  podCIDRs:
  - 10.42.0.0/24
  - 2001:cafe:42::/64
```
When I deploy canal in a dual-stack cluster, I can see the flannel infrastructure coming up correctly to support dual-stack

**Comment by manuelbuil** [comment]
If instead of using the documented config:
```
      "type": "calico",
      "ipam": {
          "type": "host-local"
          "subnet": "usePodCidr"
      },
```
I use:
```
"type": "host-local",
"ranges": [
  [{ "subnet":"10.1.2.0/24" }],
  [{ "subnet":"2001:db8:1::0/64" }} }]
],
```
Things look ok but as host-local can only guarantee the uniqueness of an IP per node, as soon as two pods share the same IP, things break.
I also tried with: `"type": "calico-ipam"` but it does not work as resources like ipamconfig are missing

**Comment by lwr20** [comment]
That's very helpful, thank you.  @caseydavenport one for you, I think?

**Comment by caseydavenport** [comment]
> I also tried with: "type": "calico-ipam" but it does not work as resources like ipamconfig are missing

Yeah, calico-ipam won't work with canal, and Calico VXLAN doesn't support IPv6 yet (there's another issue for that).

```
      "type": "calico",
      "ipam": {
          "type": "host-local"
          "subnet": "usePodCidr"
      },
```

^ So, this is going to pull addresses from spec.PodCIDR as you already discovered, which is the old single-stack field in the node. This just needs updating to support dual-stack to pull from both. Perhaps need a new value `usePodCIDRs` or something, or just change the behavior to look at the podCIDRs field.

**Comment by olljanat** [comment]
Flannel recently added dual stack support so it maybe fixed this one too? https://github.com/flannel-io/flannel/releases/tag/v0.15.0
